File: src/handoff/daemon/spawn.py
"""Headless agent spawn for failover.

When the failover policy fires, the daemon needs to launch a fresh agent in
non-interactive mode and feed it a snapshot prompt. Each agent has a
different one-shot CLI form:
    claude:  claude -p "<prompt>"
    codex:   codex exec "<prompt>"
    copilot: gh copilot suggest "<prompt>"
    cursor:  not supported (IDE-only)
"""

from __future__ import annotations

import os
import shutil
import subprocess
from pathlib import Path
from typing import Optional

from handoff.adapters import get_adapter
from handoff.paths import ca_cert_path, home_dir

NON_INTERACTIVE_ARGV: dict[str, list[str]] = {
    "claude": ["-p"],
    "codex": ["exec"],
    "copilot": ["copilot", "suggest"],  # `gh copilot suggest <prompt>`
}


def proxy_env(proxy_url: str) -> dict[str, str]:
    env = {
        "HTTP_PROXY": proxy_url,
        "HTTPS_PROXY": proxy_url,
        "http_proxy": proxy_url,
        "https_proxy": proxy_url,
    }
    ca = ca_cert_path()
    if ca.exists():
        env["SSL_CERT_FILE"] = str(ca)
        env["REQUESTS_CA_BUNDLE"] = str(ca)
        env["NODE_EXTRA_CA_CERTS"] = str(ca)
    return env


def _resolve_binary(adapter) -> Optional[str]:
    for b in adapter.binaries:
        path = shutil.which(b)
        if path:
            return path
    return None


def headless_spawn(
    kind: str,
    project_root: Path,
    prompt: str,
    proxy_url: str = "http://127.0.0.1:8080",
) -> Optional[subprocess.Popen]:
    """Spawn an agent in one-shot mode with `prompt` as input. Returns None if
    the agent has no headless mode (e.g. cursor)."""
    adapter = get_adapter(kind)
    if kind not in NON_INTERACTIVE_ARGV:
        return None
    binary = _resolve_binary(adapter)
    if binary is None:
        return None

    env = os.environ.copy()
    env.update(proxy_env(proxy_url))

    # gh copilot suggest is `gh copilot suggest <prompt>`; binary already 'gh'
    extra = NON_INTERACTIVE_ARGV[kind]
    if kind == "copilot" and Path(binary).name == "gh":
        argv = [binary, *extra, prompt]
    elif kind == "copilot":
        argv = [binary, "suggest", prompt]
    else:
        argv = [binary, *extra, prompt]

    log_path = home_dir() / f"agent-{kind}-{int(__import__('time').time())}.log"
    log_fh = open(log_path, "ab")
    return subprocess.Popen(
        argv,
        cwd=str(project_root),
        env=env,
        stdout=log_fh,
        stderr=subprocess.STDOUT,
        start_new_session=True,
    )

File: rust/crates/daemon/src/spawn.rs
//! Headless agent spawn for failover.
//!
//! `headless_spawn()` resolves the right binary and CLI args via the
//! `Adapter` trait (see `handoff_adapters::for_kind_str`), verifies the
//! binary is on PATH, then launches it with the proxy env and tee logging.

use std::collections::HashMap;
use std::path::Path;
use std::process::Stdio;

use anyhow::{anyhow, Result};
use handoff_common::home_dir;
use tokio::process::{Child, Command};

/// Returns true if `bin` resolves to an executable file anywhere in `$PATH`.
fn binary_on_path(bin: &str) -> bool {
    std::env::var_os("PATH")
        .map(|path_var| std::env::split_paths(&path_var).any(|dir| dir.join(bin).is_file()))
        .unwrap_or(false)
}

pub fn proxy_env(proxy_url: &str) -> HashMap<String, String> {
    let mut env = HashMap::new();
    env.insert("HTTP_PROXY".into(), proxy_url.into());
    env.insert("HTTPS_PROXY".into(), proxy_url.into());
    env.insert("http_proxy".into(), proxy_url.into());
    env.insert("https_proxy".into(), proxy_url.into());
    let ca = home_dir().join("ca").join("cert.pem");
    if ca.exists() {
        let s = ca.display().to_string();
        env.insert("SSL_CERT_FILE".into(), s.clone());
        env.insert("REQUESTS_CA_BUNDLE".into(), s.clone());
        env.insert("NODE_EXTRA_CA_CERTS".into(), s);
    }
    env
}

/// Spawn an agent in one-shot mode with `prompt` as input. Stdout/stderr
/// are tee'd to `~/.handoff/tee/agent-<kind>-<ts>.log`. Returns
/// `Ok(None)` for agents (e.g. cursor) that have no headless form.
pub async fn headless_spawn(
    agent_id: i64,
    kind: &str,
    project_root: &Path,
    prompt: &str,
    proxy_url: &str,
    use_worktree: bool,
) -> Result<Option<Child>> {
    let mut effective_root = project_root.to_path_buf();
    if use_worktree {
        effective_root = crate::worktree::create(project_root, agent_id)?;
    }

    // Unknown kinds (e.g. "simulated" from tests) return Ok(None) gracefully.
    let Some(adapter) = handoff_adapters::for_kind_str(kind) else {
        return Ok(None);
    };
    let args = adapter.headless_args(prompt);

    let Some(args) = args else {
        return Ok(None);
    };

    let bin = adapter.binaries()[0];

    // Verify the binary is on PATH before attempting spawn.
    if !binary_on_path(bin) {
        anyhow::bail!(
            "agent binary '{}' not found on PATH; install it before using handoff spawn",
            bin
        );
    }

    let log_path = handoff_common::tee::tee_path(agent_id);
    if let Some(parent) = log_path.parent() {
        let _ = std::fs::create_dir_all(parent);
    }
    let _ = handoff_common::tee::rotate_if_needed(&log_path);

    let log = std::fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(&log_path)
        .map_err(|e| anyhow!("opening tee log {}: {e}", log_path.display()))?;
    let log_stderr = log.try_clone()?;

    let mut cmd = Command::new(bin);
    cmd.args(&args)
        .current_dir(effective_root)
        .stdout(Stdio::from(log))
        .stderr(Stdio::from(log_stderr));
    for (k, v) in proxy_env(proxy_url) {
        cmd.env(k, v);
    }
    let child = cmd.spawn()?;
    Ok(Some(child))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn proxy_env_sets_all_keys() {
        let env = proxy_env("http://127.0.0.1:8080");
        assert!(env.contains_key("HTTPS_PROXY"));
        assert!(env.contains_key("http_proxy"));
    }

    #[test]
    fn binary_on_path_finds_sh() {
        // `sh` is present on every POSIX system.
        assert!(binary_on_path("sh"), "expected 'sh' to be on PATH");
    }

    #[test]
    fn binary_on_path_rejects_nonexistent() {
        assert!(!binary_on_path("__handoff_no_such_binary_xyz__"));
    }
}
